token.c

```diff
@@ -0,0 +1,153 @@
+/****
+ *
+ * Compile with
+ * gcc -o token -lcrypto -lcurl -Wno-deprecated-declarations token.c
+ *
+ ****/
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/time.h>
+#include <curl/curl.h>
+#include <openssl/bio.h>
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+
+struct string {
+    char *ptr;
+    size_t len;
+};
+
+static const unsigned char* encoded_key = "RmFzdGx5IFRva2VuIFRlc3Q=";
+static const unsigned long  interval    = 60;
+
+unsigned char* decode64(const unsigned char *input, int length);
+unsigned char* encode64(const unsigned char *input, int length);
+unsigned char* pack64(long value);
+void init_string(struct string *s);
+void free_string(struct string *s);
+int fetch_url(const char* url, struct string *s);
+
+int main(int argc, char** argv) {
+    struct timeval tv;
+    gettimeofday(&tv, NULL);
+    
+    const unsigned char* key    = decode64(encoded_key, strlen(encoded_key));
+    const unsigned char* number = pack64((long) tv.tv_sec / interval);
+    
+    OpenSSL_add_all_digests();
+    unsigned int md_len;
+    unsigned char* digest = HMAC(EVP_sha256(), key, strlen(key), number, 8, NULL, &md_len);
+    unsigned char* token  = encode64(digest, md_len);
+    
+    struct string response;
+    init_string(&response);
+    fetch_url("http://token.fastly.com/token", &response);
+    
+    struct string validation;
+    init_string(&validation);
+    char* url = calloc(1024, sizeof(char));
+    sprintf(url, "http://token.fastly.com?%s", token);
+    fetch_url(url, &validation);
+    
+    fprintf(stderr, "Your Token:   %s\n", token);
+    fprintf(stderr, "Fastly Token: %s\n", response.ptr);
+    fprintf(stderr, "Validation:   %s\n", validation.ptr);
+    
+    free_string(&response);
+    free_string(&validation);
+    return 0;
+}
+
+void init_string(struct string *s) {
+    s->len = 0;
+    s->ptr = malloc(s->len+1);
+    if (s->ptr == NULL) {
+        fprintf(stderr, "malloc() failed\n");
+        exit(EXIT_FAILURE);
+    }
+    s->ptr[0] = '\0';
+}
+
+void free_string(struct string *s) {
+    if (s) free(s->ptr);
+}
+
+size_t writefunc(void *ptr, size_t size, size_t nmemb, struct string *s) {
+    size_t new_len = s->len + size*nmemb;
+    s->ptr = realloc(s->ptr, new_len+1);
```

Comment:
CURL_WRITEFUNCTION documentation states:

> This function may be called with zero bytes data if the transferred file is empty. 

When the function is called with `size * nmemb == 0`, we're doing unnecessary work. Also, when the `struct string` was initialized, it was given space for the trailing `NUL` byte. The `+ 1` in the `realloc(3)` call is therefore unnecessary -- we already have space for this byte.

https://github.com/dhobsd/vstring implements a simple, self-contained string building API and supports dynamically growable strings. I'd be happy for this PR to pull vstring.h and use that for its string building needs. This would also satisfy the questions in the `TODO` above `fetch_url`.